### PR TITLE
Nim now allows modules with same name in a package

### PR DESCRIPTION
**Pull request by timotheecour**
No language I know except Nim has this restriction; lifting it just makes sense

Nim now allows modules with same name in a package, eg:
foo/bar.nim and bar.nim

this avoids having to use previous hacks like `foo/foo_bar.nim` that were there to avoid module name clashes; these were not DRY.

this enables to have `tests/stdlib/ospaths.nim` instead of `tests/stdlib/tospaths.nim` for a test of `ospaths` (**EDIT** as done for 1 test in this PR)
likewise, it allows things like:
```
json/serialization
bson/serialization

redis/types
mongodb/types
```
etc which are more natural than: `redis/redis_types`

if a user wants to import 2 modules with same basename in 1 module, he can just write:
```nim
import json/serialization as jserialization
import bson/serialization
```

## links
* https://forum.nim-lang.org/t/2937
* https://gitter.im/nim-lang/Nim?at=5b6e9829988005174ec35f95

## NOTE
* suppose we have:
`/foo/bar/jesterRepo/util/baz.nim` 
with `/foo/bar/jesterRepo/jester.nimble` 
it generates:
`jester.util` as package name **EDIT** 
(ie using `getPackageName` which finds the first dir which contains a nimble file)

**EDIT**  The following may be a bit surprising but are result of current placement of files in Nim:
```
stdlib.std.sha1
stdlib.pure.collections.intsets
compiler.compiler.prefixmatches # because compiler.nimble is under Nim/compiler.nimble instead of Nim/compiler/compiler.nimble => to be fixed in PR https://github.com/nim-lang/Nim/pull/8636
```

* **EDIT**  it generates package parents hierachically, exactly as follows:
`module ospaths` -> `package stdlib.pure` (note: fully qualified) => `package stdlib`

* **EDIT** files like these will be generated in nimcache:
```
stdlib.pure.bitops.c
stdlib.pure.math.c
stdlib.system.c
timn.bugs.exp_module_nonuniq.t01.c
```

* **EDIT** example of things it generates during compilation (with `--hint[CC]=on`):
```
CC: stdlib.pure.collections.rtarrays
CC: stdlib.std.sha1
CC: compiler.compiler.prefixmatches
CC: stdlib.pure.concurrency.cpuinfo
CC: stdlib.impure.nre.private.util
CC: stdlib.impure.nre
CC: stdlib.deprecated.pure.asyncio
CC: stdlib.system
```

## TODO
- [x] **EDIT** add tests => tests/modules/tnotuniquename1.nim
- [ ] **EDIT** not yet sure how to handle this 1 particular line: `registerCallback c, "stdlib.*.staticWalkDir", proc (a: VmArgs) {.nimcall.} =` ; @araq ?

## TODO (future PR)
- [ ] move Nim/compiler.nimble to Nim/compiler/compiler.nimble => PR https://github.com/nim-lang/Nim/pull/8636
- [ ] rename stdlib.nimble to std.nimble ? (would make more sense with `import std/random`) => PR https://github.com/nim-lang/Nim/pull/8637

## depends on:
* https://github.com/nim-lang/Nim/pull/8631

## previous versions of this PR
a previous version of this PR was using a simpler scheme:
`/foo/bar/jester/util/baz.nim`  was mapped to `@foo@bar@jester@util@` as package name ; this caused issues, eg with `registerCallback` and other tests, and made packages depend on location of nimble packages on filesystem, which is not desirable

then I made changes to use  `@jester@util@` instead of `@foo@bar@jester@util@` (ie using `getPackageName` to find package root)

then I made changes to use `jester.util` instead of `@jester@util@` which looks more natural

### CODE REVIEW

**Comment by alehander42** [comment]
It would be cool if this limitation is lifted, I also find that it just leads to hacky workarounds

**Comment by Araq** [comment]
FYI this was actually on my TODO list too for the upcoming release.

File: compiler/packagehandling.nim [hunk 1]
```diff
@@ -18,7 +18,9 @@ iterator myParentDirs(p: string): string =
 proc resetPackageCache*(conf: ConfigRef) =
   conf.packageCache = newPackageCache()
 
-proc getPackageName*(conf: ConfigRef; path: string): string =
+when false:
+ proc getPackageName*(conf: ConfigRef; path: string): string =
```
Comment (euantorano) [inline compiler/packagehandling.nim]: Indenting here looks wrong, it should be 2 spaces where this appears to just be 1...

**Comment by Clyybber** [comment]
@Araq Is that TODO list public?

File: compiler/packagehandling.nim (hunk 1 above)
Comment (dom96) [reply compiler/packagehandling.nim]: In fact, why not just remove this code instead of disabling it?

**Comment by dom96** [comment]
Minor point, by why ``@``?

**Comment by Araq** [comment]
@Clyybber With this feature implemented and my ``araq-misc`` branch merged, it'll be empty. (The other items are all on github.)

**Comment by Araq** [comment]
> Minor point, by why @?

That won't stay, it's not allowed on Windows anyway.

File: compiler/packagehandling.nim (hunk 1 above)
Comment (timotheecour) [reply compiler/packagehandling.nim]: fixed

**Comment by timotheecour** [comment]
@araq @dom96 
> Minor point, by why @?
> That won't stay, it's not allowed on Windows anyway.

Actually `@` is allowed for filenames on windows, see https://stackoverflow.com/a/31976060/1426932 as well as green tests in https://github.com/nim-lang/Nim/pull/8629 which was a previous version of this PR using `@` instead (I also tried to create `foo@bar`on windows command prompt and it works)

However, I agree `@` was not the best choice and now used `.` in this PR which fits better with rest of Nim.

PTAL

**Comment by dom96** [comment]
Why don't you just use ``_``?

**Comment by timotheecour** [comment]
@araq @dom96 PTAL: tests are green

**Comment by GULPF** [comment]
How will this work with the docgen? It currently puts every HTML file in the same directory, which will cause conflicts if there are duplicate module names.

**Comment by dom96** [comment]
> I think the docgen must be changed as well, it currently puts every HTML file in the same directory which will cause conflicts if there are duplicate module names.

I disagree. Worst case is the user will have to write ``-o:my_module2.html``.

**Comment by GULPF** [comment]
> I disagree. Worst case is the user will have to write -o:my_module2.html.

Using `-o` with a different file name means that module links that are generated by the docgen wont work. [These are the links I'm referring to](https://nim-lang.org/docs/times.html#6). Besides, writing the `nim doc` commands for individual modules isn't really feasibly in large projects (like Nims stdlib).

**Comment by Clyybber** [comment]
@timotheecour @Araq Will we get this merged before 0.19? If the comments are addressed at least?

**Comment by Araq** [comment]
Unlikely, sorry. v0.19 is overdue and bugs like https://github.com/nim-lang/Nim/issues/7854 take up all resources. :-)

**Comment by Araq** [comment]
As I said.

> This is hard to follow and I don't understand why it's necessary. For the frontend you can just disable this check. For the backend you can replicate the Path structure or detect the part of the path required for disambiguation. Or you hash the full path and append it to the filename. These ideas should all lead to simpler code that doesn't touch so many parts of the compiler.
